modelling/src/main/java/org/axonframework/modelling/saga/AnnotatedSaga.java

```diff
@@ -131,9 +131,7 @@ private Object handle(MessageHandlingMember<? super T> handler, EventMessage<?>
     }
 
     private Boolean isEndingHandler(MessageHandlingMember<? super T> handler) {
-        return handler.unwrap(SagaMethodMessageHandlingMember.class)
-                      .map(SagaMethodMessageHandlingMember::isEndingHandler)
-                      .orElse(false);
+        return handler.hasAnnotation(EndSaga.class);
```

Comment:
for backwards compatibility, wouldn't it be better to check for the `SagaMethodMessageHandlingMember::isEndingHandler` and in the `orElse` do the `handler.hasAnnotation(EndSaga.class)`.
I also think that the `isEndingHandler` is invoked for each Saga Handler invocation. I'm wondering if it wouldn't be better to wrap an Event Handler annotated with `@EndSaga` with the behavior to end the Saga using `SagaLifecycle.end()`

Reply:
Just looked at the code. It's better to deprecate the `isEndlingHandler` method and create a MessageHandlerWrapper that reacts on the EndSaga annotation. That way, there is no need to inspect annotations for each call, but rather only during initialization.

Reply:
Was already in doubt whether I could "get away" with this simple, albeit hacky, solution.

Just added a dedicated handler enhancer definition and handling member to achieve the same. 
Let me know what you think.